File: 0049-group-anagrams/0049-group-anagrams.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> hash;

        for(string s : strs) {
            string key = s;
            sort(key.begin(), key.end()); // same anagrams will have same sorted string
            hash[key].push_back(s);
        }

        vector<vector<string>> result;
        for(auto& it : hash) {
            result.push_back(it.second);
        }

        return result;
    }
};

File: 0049-group-anagrams/0049-group-anagrams.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<>();

        for(String str : strs) {
            String s = sortS(str);

            if(!map.containsKey(s)) {
                map.put(s, new ArrayList<>());
            }

            map.get(s).add(str);
        }

        List<List<String>> res = new ArrayList<>();

        for(String key : map.keySet()) {
            res.add(map.get(key));
        }

        return res;

    }

    private String sortS(String str) {
        char[] arr = str.toCharArray();
        Arrays.sort(arr);

        String sorted = new String(arr);
        return sorted;
    }
}